crates/engine/tree/src/engine.rs

```diff
@@ -0,0 +1,223 @@
+/// chain).
+///
+/// It advances the chain based on received engine API requests:
+///
+/// - `on_new_payload`: Executes the payload and inserts it into the tree. These are allowed to be
+///   processed concurrently.
+/// - `on_forkchoice_updated`: Updates the fork choice based on the new head. These require write
+///   access to the database and are skipped if the handler can't acquire exclusive access to the
+///   database.
+///
+/// The [`EngineApiTreeHandler`] is used to execute the incoming payloads, storing them in a tree
+/// structure and committing new chains to the database on fork choice updates.
+///
+/// In case required blocks are missing, the handler will request them from the network, by emitting
+/// a download request upstream.
+#[derive(Debug)]
+pub struct EngineApiRequestHandler<T>
+where
+    T: EngineApiTreeHandler,
+{
+    /// The state of the top level orchestrator.
+    orchestrator_state: OrchestratorState,
+    /// Needs to keep track of requested state changes by the orchestrator.
+    state: (),
+    /// Currently in progress payload requests.
+    pending_payloads: Vec<()>,
+    /// Currently in progress fork choice updates.
+    pending_fcu: Option<()>,
+    /// Next FCU to process
+    next_fcu: VecDeque<()>,
+    /// Manages the tree
+    tree_handler: T,
+    /// Events to yield.
+    buffered_events: VecDeque<()>,
+}
+
+impl<T> EngineRequestHandler for EngineApiRequestHandler<T>
+where
+    T: EngineApiTreeHandler,
+{
+    type Request = BeaconEngineMessage<T::Engine>;
+
+    fn on_event(&mut self, event: FromEngine<Self::Request>) {
+        // TODO check if we're currently syncing, or mutable access is currently held by the
+        // orchestrator then we respond with SYNCING or delay forkchoice updates.  otherwise
+        // we tell the tree to handle the requests, but we likely still need to tell the handler
+        // about the stuff while write access is unavailable.
+
+        // TODO: should this type spawn the jobs and have access to tree internals or should this be
+        // entirely handled by the tree handler? basically
+
+        /*
+           let (tx, req) = event;
+           let handler = self.tree_handler.clone();
+           spawn(move {
+               let resp = handler.on_new_payload(req);
+               tx.send(resp).unwrap();
+           });
+        */
+        // here the handler must contain shareable state internally
+
+        // or
+
+        /*
+         let fut = handler.handle(event);
+         self.pending.push(fut);
+        */
+
+        // the latter would give the handler more control over the execution of the requests, with
+        // this model the logic of this type and the tree handler is very similar
+
+        todo!()
+    }
+
+    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<RequestHandlerEvent> {
+        // advance tree tasks, trigger
+        todo!()
+    }
```

Comment:
let's use proper impl of either `Future` or `Stream` trait, and not custom poll functions. probably `Stream` is best fit for this type, let's not implement "endless future"s. the `Stream` should be implemented like a generator, which means it should not contain loops. if you need to put a loop in it, pls use the `metered_poll_nested_stream_with_budget` macro.

Reply:
these are incompatible with the future or stream interface, rolling our own traits is a lot easier

Reply:
makes much easier to integrate reth though in other software/for users to build custom components built into reth, if we use rust's standard for async interfaces @mattsse